File: mypalclara/services/scheduler/tools.py
"""Tool definitions for scheduler management."""

from __future__ import annotations

import uuid
from datetime import UTC, datetime
from typing import Any

from .models import ScheduledTask, TaskStatus, TaskType
from .scheduler import Scheduler

SCHEDULER_TOOLS: list[dict[str, Any]] = [
    {
        "type": "function",
        "function": {
            "name": "schedule_task",
            "description": (
                "Schedule a one-shot task to run at a specific time. "
                "The task dispatches a prompt through the gateway pipeline."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "prompt": {
                        "type": "string",
                        "description": "The prompt to dispatch when the task fires.",
                    },
                    "run_at": {
                        "type": "string",
                        "description": "ISO 8601 datetime for when to run (e.g. '2026-03-20T15:00:00Z').",
                    },
                    "description": {
                        "type": "string",
                        "description": "Human-readable description of the task.",
                    },
                },
                "required": ["prompt", "run_at"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "schedule_cron",
            "description": (
                "Schedule a recurring task using a cron expression. "
                "The task dispatches a prompt through the gateway pipeline on each fire."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "prompt": {
                        "type": "string",
                        "description": "The prompt to dispatch on each fire.",
                    },
                    "cron_expr": {
                        "type": "string",
                        "description": "Standard 5-field cron expression (e.g. '*/30 * * * *' for every 30 min).",
                    },
                    "description": {
                        "type": "string",
                        "description": "Human-readable description of the recurring task.",
                    },
                },
                "required": ["prompt", "cron_expr"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "list_scheduled_tasks",
            "description": "List all scheduled tasks (both one-shot and recurring).",
            "parameters": {
                "type": "object",
                "properties": {},
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "cancel_scheduled_task",
            "description": "Cancel and remove a scheduled task by its ID.",
            "parameters": {
                "type": "object",
                "properties": {
                    "task_id": {
                        "type": "string",
                        "description": "The ID of the task to cancel.",
                    },
                },
                "required": ["task_id"],
            },
        },
    },
]


def handle_scheduler_tool(
    tool_name: str,
    args: dict[str, Any],
    scheduler: Scheduler,
    user_id: str,
    channel_id: str,
) -> str:
    """Dispatch a scheduler tool call and return a human-readable result.

    Args:
        tool_name: One of the scheduler tool names.
        args: Tool arguments.
        scheduler: The active Scheduler instance.
        user_id: The calling user's ID.
        channel_id: The channel where the tool was invoked.

    Returns:
        A string result message.
    """
    if tool_name == "schedule_task":
        return _handle_schedule_task(args, scheduler, user_id, channel_id)
    if tool_name == "schedule_cron":
        return _handle_schedule_cron(args, scheduler, user_id, channel_id)
    if tool_name == "list_scheduled_tasks":
        return _handle_list(scheduler, user_id)
    if tool_name == "cancel_scheduled_task":
        return _handle_cancel(args, scheduler)
    return f"Unknown scheduler tool: {tool_name}"


def _handle_schedule_task(
    args: dict[str, Any],
    scheduler: Scheduler,
    user_id: str,
    channel_id: str,
) -> str:
    prompt = args["prompt"]
    run_at_str = args["run_at"]
    description = args.get("description", "")

    try:
        run_at = datetime.fromisoformat(run_at_str)
        if run_at.tzinfo is None:
            run_at = run_at.replace(tzinfo=UTC)
    except ValueError:
        return f"Invalid datetime format: {run_at_str!r}. Use ISO 8601 (e.g. '2026-03-20T15:00:00Z')."

    task_id = f"task-{uuid.uuid4().hex[:8]}"
    task = ScheduledTask(
        id=task_id,
        type=TaskType.ONE_SHOT,
        prompt=prompt,
        user_id=user_id,
        channel_id=channel_id,
        run_at=run_at,
        description=description,
    )
    scheduler.add_task(task)
    return f"Scheduled one-shot task {task_id} for {run_at.isoformat()}: {description or prompt[:60]}"


def _handle_schedule_cron(
    args: dict[str, Any],
    scheduler: Scheduler,
    user_id: str,
    channel_id: str,
) -> str:
    prompt = args["prompt"]
    cron_expr = args["cron_expr"]
    description = args.get("description", "")

    task_id = f"cron-{uuid.uuid4().hex[:8]}"
    task = ScheduledTask(
        id=task_id,
        type=TaskType.CRON,
        prompt=prompt,
        user_id=user_id,
        channel_id=channel_id,
        cron_expr=cron_expr,
        description=description,
    )
    scheduler.add_task(task)
    return f"Scheduled recurring task {task_id} ({cron_expr}): {description or prompt[:60]}"


def _handle_list(scheduler: Scheduler, user_id: str) -> str:
    tasks = scheduler.list_tasks(user_id=user_id)
    if not tasks:
        return "No scheduled tasks."
    lines = []
    for t in tasks:
        kind = "cron" if t.type == TaskType.CRON else "one-shot"
        schedule = t.cron_expr if t.type == TaskType.CRON else (t.run_at.isoformat() if t.run_at else "unset")
        desc = t.description or t.prompt[:40]
        lines.append(f"- {t.id} [{kind}, {t.status.value}] {schedule} — {desc}")
    return "\n".join(lines)


def _handle_cancel(args: dict[str, Any], scheduler: Scheduler) -> str:
    task_id = args["task_id"]
    if scheduler.remove_task(task_id):
        return f"Cancelled task {task_id}."
    return f"No task found with ID {task_id}."

File: go/internal/services/scheduler/tools.go
package scheduler

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/BangRocket/mypalclara/go/internal/llm"
)

// ToolSchemas returns LLM tool schemas for scheduler operations.
func ToolSchemas() []llm.ToolSchema {
	return []llm.ToolSchema{
		{
			Name:        "schedule_task",
			Description: "Schedule a one-shot task to run at a specific time.",
			Parameters: map[string]any{
				"type": "object",
				"properties": map[string]any{
					"prompt": map[string]any{
						"type":        "string",
						"description": "The prompt/instruction to execute when the task fires.",
					},
					"run_at": map[string]any{
						"type":        "string",
						"description": "When to run the task (RFC 3339 format, e.g. 2025-01-15T14:30:00Z).",
					},
					"description": map[string]any{
						"type":        "string",
						"description": "Human-readable description of the task.",
					},
				},
				"required": []string{"prompt", "run_at"},
			},
		},
		{
			Name:        "schedule_cron",
			Description: "Schedule a recurring task using a cron expression.",
			Parameters: map[string]any{
				"type": "object",
				"properties": map[string]any{
					"prompt": map[string]any{
						"type":        "string",
						"description": "The prompt/instruction to execute on each run.",
					},
					"cron_expr": map[string]any{
						"type":        "string",
						"description": "Cron expression (e.g. '0 9 * * *' for daily at 9am).",
					},
					"description": map[string]any{
						"type":        "string",
						"description": "Human-readable description of the recurring task.",
					},
				},
				"required": []string{"prompt", "cron_expr"},
			},
		},
		{
			Name:        "list_scheduled_tasks",
			Description: "List all scheduled tasks for the current user.",
			Parameters: map[string]any{
				"type":       "object",
				"properties": map[string]any{},
			},
		},
		{
			Name:        "cancel_scheduled_task",
			Description: "Cancel a scheduled task by ID.",
			Parameters: map[string]any{
				"type": "object",
				"properties": map[string]any{
					"task_id": map[string]any{
						"type":        "string",
						"description": "The ID of the task to cancel.",
					},
				},
				"required": []string{"task_id"},
			},
		},
	}
}

// HandleToolCall dispatches a scheduler tool call to the appropriate handler.
// Returns the result string and any error.
func HandleToolCall(s *Scheduler, ctx context.Context, toolName string, args map[string]any, userID string) (string, error) {
	switch toolName {
	case "schedule_task":
		return handleScheduleTask(s, args, userID)
	case "schedule_cron":
		return handleScheduleCron(s, args, userID)
	case "list_scheduled_tasks":
		return handleListTasks(s, userID)
	case "cancel_scheduled_task":
		return handleCancelTask(s, args)
	default:
		return "", fmt.Errorf("unknown scheduler tool: %s", toolName)
	}
}

func handleScheduleTask(s *Scheduler, args map[string]any, userID string) (string, error) {
	prompt, _ := args["prompt"].(string)
	runAtStr, _ := args["run_at"].(string)
	desc, _ := args["description"].(string)

	if prompt == "" {
		return "", fmt.Errorf("prompt is required")
	}
	if runAtStr == "" {
		return "", fmt.Errorf("run_at is required")
	}

	runAt, err := time.Parse(time.RFC3339, runAtStr)
	if err != nil {
		return "", fmt.Errorf("invalid run_at format (expected RFC 3339): %w", err)
	}

	task := &ScheduledTask{
		Type:        TaskTypeOneShot,
		Prompt:      prompt,
		UserID:      userID,
		RunAt:       &runAt,
		Description: desc,
	}
	s.AddTask(task)

	return fmt.Sprintf("Scheduled task %s to run at %s", task.ID, runAt.Format(time.RFC3339)), nil
}

func handleScheduleCron(s *Scheduler, args map[string]any, userID string) (string, error) {
	prompt, _ := args["prompt"].(string)
	cronExpr, _ := args["cron_expr"].(string)
	desc, _ := args["description"].(string)

	if prompt == "" {
		return "", fmt.Errorf("prompt is required")
	}
	if cronExpr == "" {
		return "", fmt.Errorf("cron_expr is required")
	}

	task := &ScheduledTask{
		Type:        TaskTypeCron,
		Prompt:      prompt,
		UserID:      userID,
		CronExpr:    cronExpr,
		Description: desc,
	}
	s.AddTask(task)

	return fmt.Sprintf("Scheduled cron task %s with expression %q", task.ID, cronExpr), nil
}

func handleListTasks(s *Scheduler, userID string) (string, error) {
	tasks := s.ListTasks(userID)
	if len(tasks) == 0 {
		return "No scheduled tasks.", nil
	}

	var sb strings.Builder
	for _, t := range tasks {
		sb.WriteString(fmt.Sprintf("- [%s] %s (%s) status=%s", t.ID, t.Description, t.Type, t.Status))
		if t.RunAt != nil {
			sb.WriteString(fmt.Sprintf(" run_at=%s", t.RunAt.Format(time.RFC3339)))
		}
		if t.CronExpr != "" {
			sb.WriteString(fmt.Sprintf(" cron=%q", t.CronExpr))
		}
		sb.WriteString("\n")
	}

	return sb.String(), nil
}

func handleCancelTask(s *Scheduler, args map[string]any) (string, error) {
	taskID, _ := args["task_id"].(string)
	if taskID == "" {
		return "", fmt.Errorf("task_id is required")
	}

	if s.RemoveTask(taskID) {
		return fmt.Sprintf("Cancelled task %s", taskID), nil
	}
	return "", fmt.Errorf("task %s not found", taskID)
}
